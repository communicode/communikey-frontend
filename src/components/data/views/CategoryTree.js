import React from "react";
import PropTypes from "prop-types";
import {PropTypes as MobXPropTypes} from "mobx-react";
import {Icon, Tree} from "antd";
import "antd/lib/icon/style/css";
import "antd/lib/tree/style/index.less";
import "./CategoryTree.less";

/**
 * A tree for categories.
 *
 * @author sgreb@communicode.de
 * @since 0.8.0
 */
class CategoryTree extends React.Component {
  constructor(props) {
    super(props);
  }

  /**
   * Recursivly generates all tree nodes.
   *
   * @param categories
   */
  generateTreeNodes = categories => categories.map(category => {
    if (category.children.length > 0) {
      return (
        <Tree.TreeNode key={category.id} title={<span><Icon type="folder"/> {category.name}</span>} category={category}>
          {this.generateTreeNodes(category.children)}
        </Tree.TreeNode>
      );
    }
    return <Tree.TreeNode key={category.id} title={category.name} category={category}/>;
  });

  render() {
    const {categories, expandedKeys, onExpand, onSelect, selectedKeys, ...treeProps} = this.props;

    return (
      <Tree
        {...treeProps}
        expandedKeys={expandedKeys}
        onExpand={onExpand}
        onSelect={onSelect}
        selectedKeys={selectedKeys}
      >
        {this.generateTreeNodes(categories)}
      </Tree>
    );
  }
}

CategoryTree.propTypes = {
  /**
   * The categories.
   */
  categories: MobXPropTypes.observableArray.isRequired,

  /**
   * The currently expanded keys.
   */
  expandedKeys: PropTypes.array.isRequired,

  /**
   * Callback function to handle expand events.
   */
  onExpand: PropTypes.func.isRequired,

  /**
   * Callback function to handle selection events.
   */
  onSelect: PropTypes.func.isRequired,

  /**
   * The currently selected keys.
   */
  selectedKeys: PropTypes.array.isRequired
};

export default CategoryTree;