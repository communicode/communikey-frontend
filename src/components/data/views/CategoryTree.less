.category-tree {
  &.ant-tree li span[draggable],
  &.ant-tree li span[draggable="true"] {
    border-top: none;
    border-bottom: none;
    margin-top: 0;
  }
}