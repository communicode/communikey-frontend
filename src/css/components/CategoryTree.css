.categoryTree {
  float: left !important;
  height: 100%;
  width: 250px !important;
  overflow: auto;
  display: inline-block !important;
}

.accordion {
  padding-left: 15px !important;
  margin-top: 0 !important;
}

.accordion .content {
  padding-top: 0 !important;
}
