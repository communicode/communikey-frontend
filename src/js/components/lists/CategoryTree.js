import React, {Component} from "react";
import {PropTypes} from "mobx-react";
import {Accordion, Icon} from "semantic-ui-react";

/**
 * A tree for categories to be wrapped into a {@link Menu} Semantic UI component.
 *
 * @author sgreb@communicode.de
 * @since 0.6.0
 */
class CategoryTree extends Component {
  constructor(props) {
    super(props);
    this.state = {
      /**
       * The currently selected category.
       *
       * @type {object}
       * @default null
       */
      selectedCategory: null
    }
  }

  /**
   * Handles the event when the category name gets clicked.
   * Fires the {@code onCategorySelect()} prop callback function.
   *
   * @param {object} category - The category of which the title has been clicked
   */
  handleTitleClick = (category) => {
    this.setState({selectedCategory: category});
    this.props.onCategorySelect(category);
  };

  render() {
    return (
      <div>
        {this.props.categories.map(category => (
            <Accordion fluid={true} key={category.id} className="link item" onTitleClick={() => this.handleTitleClick(category)}>
              <Accordion.Title>
                <Icon name="dropdown"/>
                {category.name}
              </Accordion.Title>
              <Accordion.Content>
                {category.children &&
                <CategoryTree
                  categories={category.children}
                  onTitleClick={() => this.handleTitleClick(category)}
                  onCategorySelect={this.props.onCategorySelect}
                />}
              </Accordion.Content>
            </Accordion>
          )
        )}
      </div>
    )
  }
}

CategoryTree.propTypes = {
  /**
   * @type {ObservableArray} categories - The observable array of categories
   */
  categories: PropTypes.observableArray.isRequired
};

export default CategoryTree;