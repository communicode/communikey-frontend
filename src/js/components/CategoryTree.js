import React, { Component } from "react";
import {inject, observer} from "mobx-react";
import {Accordion, Icon, Segment} from "semantic-ui-react";
import KeyCardView from "./KeyCardView";
import "../../css/components/CategoryTree.css"

@inject("categoryStore") @observer
class CategoryTree extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedCategory: ""
    }
  }

  changeMyState(category) {
    //TODO: Remove comments
    // console.log(this.state);
    // this.state.selectedCategory = category;
    // this.forceUpdate();
    this.setState({selectedCategory: category});
  }

  componentDidMount() {
    this.props.categoryStore.fetchCategories();
  }

  render() {
    const { activeIndex, selectedCategory } = this.state
    let handleClick = (category, index) => this.changeMyState(category)

    function CategoryList(props) {
      if(props.node !== undefined && props.node.length > 0) {
        let rows = [];
        props.node.map(function(category) {
          rows.push(
            <Accordion fluid activeIndex={0} key={category.id}>
              <Accordion.Title onClick={() => handleClick(category)}>
                <Icon name='triangle right' />
                {category.name}
              </Accordion.Title>
              <Accordion.Content>
                <CategoryList node={category.children} />
              </Accordion.Content>
            </Accordion>
          );
        })
        return <tbody>{rows}</tbody>;
      } else return null
    }

    return (
      <div>
        <Segment class="categoryTree">
          <CategoryList node={this.props.categoryStore.categories} />
        </Segment>
        <KeyCardView category={this.state.selectedCategory}/>
      </div>
      )
    }
  }

  export default CategoryTree
