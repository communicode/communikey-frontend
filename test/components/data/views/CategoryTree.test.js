import React from "react";
import {shallow} from "enzyme";
import toJson from 'enzyme-to-json';
import CategoryTree from "../../../../src/components/data/views/CategoryTree";
import {categoryStore, emptyFunction} from "../../../__mockData__/mockUtil";

describe('<CategoryTree>', () => {
  test('should render correctly with sample data', () => {
    const wrapper = shallow(<CategoryTree categories={categoryStore.categories} onSelect={emptyFunction}/>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});