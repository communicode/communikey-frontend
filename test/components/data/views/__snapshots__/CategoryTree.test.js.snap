// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryTree> should render correctly with sample data 1`] = `
<Tree
  checkable={false}
  draggable={false}
  onSelect={[Function]}
  openAnimation={
    Object {
      "appear": [Function],
      "enter": [Function],
      "leave": [Function],
    }
  }
  prefixCls="ant-tree"
  showIcon={false}
>
  <TreeNode
    category={
      Object {
        "children": Array [
          5,
        ],
        "createdBy": "root",
        "createdDate": "2017-06-30T11:13:46.691+02:00",
        "creator": 1,
        "groups": Array [],
        "id": 4,
        "keys": Array [],
        "lastModifiedBy": "root",
        "lastModifiedDate": "2017-06-30T11:13:46.696+02:00",
        "name": "intern",
        "parent": null,
        "responsible": 1,
        "treeLevel": 0,
      }
    }
    title={
      <span>
        <Icon
          type="folder"
        />
         
        intern
      </span>
    }
  >
    <TreeNode
      category={
        Object {
          "children": Array [],
          "createdBy": "root",
          "createdDate": "2017-06-30T11:13:53.821+02:00",
          "creator": 1,
          "groups": Array [],
          "id": 5,
          "keys": Array [
            3,
          ],
          "lastModifiedBy": "root",
          "lastModifiedDate": "2017-06-30T11:13:55.992+02:00",
          "name": "projektron bcs",
          "parent": 4,
          "responsible": 1,
          "treeLevel": 1,
        }
      }
      title="projektron bcs"
    />
  </TreeNode>
  <TreeNode
    category={
      Object {
        "children": Array [
          2,
        ],
        "createdBy": "root",
        "createdDate": "2017-06-30T11:13:33.274+02:00",
        "creator": 1,
        "groups": Array [],
        "id": 1,
        "keys": Array [],
        "lastModifiedBy": "root",
        "lastModifiedDate": "2017-06-30T11:13:33.282+02:00",
        "name": "vorwerk",
        "parent": null,
        "responsible": 1,
        "treeLevel": 0,
      }
    }
    title={
      <span>
        <Icon
          type="folder"
        />
         
        vorwerk
      </span>
    }
  >
    <TreeNode
      category={
        Object {
          "children": Array [
            3,
          ],
          "createdBy": "root",
          "createdDate": "2017-06-30T11:13:37.557+02:00",
          "creator": 1,
          "groups": Array [],
          "id": 2,
          "keys": Array [
            1,
          ],
          "lastModifiedBy": "root",
          "lastModifiedDate": "2017-06-30T11:14:04.707+02:00",
          "name": "servers",
          "parent": 1,
          "responsible": 1,
          "treeLevel": 1,
        }
      }
      title={
        <span>
          <Icon
            type="folder"
          />
           
          servers
        </span>
      }
    >
      <TreeNode
        category={
          Object {
            "children": Array [],
            "createdBy": "root",
            "createdDate": "2017-06-30T11:13:41.497+02:00",
            "creator": 1,
            "groups": Array [],
            "id": 3,
            "keys": Array [
              2,
            ],
            "lastModifiedBy": "root",
            "lastModifiedDate": "2017-06-30T11:14:01.09+02:00",
            "name": "ssh",
            "parent": 2,
            "responsible": 1,
            "treeLevel": 2,
          }
        }
        title="ssh"
      />
    </TreeNode>
  </TreeNode>
</Tree>
`;
